File: engine/negative_controls/xxe.py
"""Negative control: "xxe" keyword family (Python).

defusedxml parsing only — entity resolution structurally disabled by
the library. No lxml parse call, no entity-resolution opt-in, no DTD
loading. The unsafe-shape pattern must not match any line of this
fixture.
"""

from defusedxml.ElementTree import fromstring as safe_fromstring


def parse_manifest(text):
    # defusedxml forbids entity expansion and external DTD fetches by
    # construction — the safe idiom the unsafe-shape pattern must not
    # match.
    return safe_fromstring(text)

File: engine/negative_controls/xxe.c
/* Negative control: "xxe" keyword family (C).
 *
 * libxml2 parse with network access disabled and entity substitution
 * left at its safe default (no opt-in flag, no global substitution
 * toggle). The unsafe-shape pattern must not match.
 */
#include <libxml/parser.h>

xmlDocPtr parse_manifest(const char *path)
{
    /* NONET blocks external fetches; entity substitution stays off
     * because the opt-in parse flag is deliberately absent. */
    return xmlReadFile(path, NULL, XML_PARSE_NONET);
}
